### PR TITLE
add mkldnn prop_kind phase for inference-only case to pooling and activations

### PR DESCRIPTION
**Pull request by sfraczek**
This PR adds is_test flag to two operators: pooling and activations and fixes convolution to use it also.
It adds a capi pass that sets is_test to true.
It adds a test for that capi pass.
It adds a transpiler pass with the same functionality.

### CODE REVIEW

**Comment by sfraczek** [comment]
I need to update the test because it did not catch one error.

File: paddle/fluid/inference/analysis/analyzer.h [hunk 1]
```diff
@@ -83,6 +83,7 @@ class Analyzer : public OrderedRegistry<PassManager> {
       "conv_bias_mkldnn_fuse_pass",             //
       "conv_relu_mkldnn_fuse_pass",             //
       "conv_elementwise_add_mkldnn_fuse_pass",  //
+      "mkldnn_correct_test_phase_pass",         //
```
Comment (luotao1) [inline paddle/fluid/inference/analysis/analyzer.h]: Should `mkldnn_correct_test_phase_pass` be put before `depthwise_conv_mkldnn_pass` ?

**Review by luotao1 — COMMENTED** [review]
How about add description of `is_test` attribute of `sequence_pool` https://github.com/PaddlePaddle/Paddle/pull/14053#discussion_r229558205 if you are convenient?

File: paddle/fluid/inference/analysis/analyzer.h (hunk 1 above)
Comment (sfraczek) [reply paddle/fluid/inference/analysis/analyzer.h]: `mkldnn_correct_test_phase_pass` can be put anywhere. The later it is called, the less layers there are left and thus the less work it has to do.

Comment (luotao1) [reply paddle/fluid/inference/analysis/analyzer.h]: Got it

**Comment by sfraczek** [comment]
Sure @luotao1 , I will also add the description in `sequence_pool` and other.

**Comment by sfraczek** [comment]
The `open_files_op.cc` and `while_op.cc` operator seems to have `is_test` attribute serving a different purpose. This could be a problem if my pass sets this to true when it shouldn't. @luotao1 @mrysztow @kbinias @jczaja @wojtuss , what should we do?
I will make the pass set this attribute only for layers that use mkldnn. This way it won't modify the open_files_op's or while_op's attribute, but still it has the same name which might be confusing.

File: paddle/fluid/framework/ir/mkldnn_correct_test_phase_pass.cc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      "elu",         "floor",        "cos",
+      "sin",         "round",        "reciprocal",
+      "hard_shrink", "hard_sigmoid", "relu6",
+      "soft_relu",   "swish",        "thresholded_relu",
+      "log",         "square",       "softplus",
+      "softsign"};
+  for (const Node* n : graph->Nodes()) {
+    if (n->IsOp()) {
+      auto* op = n->Op();
+      if (op->HasAttr("use_mkldnn") && op->GetAttr("use_mkldnn")) {
```
Comment (luotao1) [inline paddle/fluid/framework/ir/mkldnn_correct_test_phase_pass.cc]: why need `op->GetAttr("use_mkldnn")`?

Comment (sfraczek) [reply paddle/fluid/framework/ir/mkldnn_correct_test_phase_pass.cc]: It is my mistake. Non MKLDNN layers could also be optimized for inference, right? 
I am not sure if `is_test` flag has the same use in all layers especially in the below 3 layers.

1. `fake_quantize_op.cc` has `AddAttr<bool>("is_test", "").SetDefault(false);`. 
2. `open_files_op.cc` has  ` AddAttr<bool>("is_test", "Used for testing data.").SetDefault(false);`. 
3. `while_op.cc`. has `AddAttr<bool>("is_test", "True if in test phase.").SetDefault(false);`

They seem to be used the same way but can you confirm?

File: paddle/fluid/inference/analysis/analyzer.h (hunk 1 above)
Comment (sfraczek) [reply paddle/fluid/inference/analysis/analyzer.h]: I actually put it outside the ifdef scope now after renaming it.

File: paddle/fluid/framework/ir/mkldnn_correct_test_phase_pass.cc (hunk 2 above)
Comment (sfraczek) [reply paddle/fluid/framework/ir/mkldnn_correct_test_phase_pass.cc]: I will assume they are the same.

**Comment by tpatejko** [comment]
@sfraczek Have you checked if there are any MKLDNN reorder operations of convolution weights? If I'm not mistaken, `is_test` flag in the MKLDNN convolution operator is used for checking if weights require reorder to MKLDNN format.

Was this PR tested against any performance regressions?

File: paddle/fluid/operators/activation_mkldnn_op.cc [hunk 3]
```diff
@@ -115,11 +119,13 @@ void eltwise_forward(const framework::ExecutionContext &ctx,
...
 
+  bool is_test = ctx.Attr<bool>("is_test");
+
   // save input data and layout to be referred in backward path
   auto p_src_data = std::make_shared<const T *>(x_data);
-  dev_ctx.SetBlob(key_src_data, p_src_data);
+  if (!is_test) dev_ctx.SetBlob(key_src_data, p_src_data);
   auto p_src_layout = std::make_shared<memory::format>(src_format);
-  dev_ctx.SetBlob(key_src_layout, p_src_layout);
+  if (!is_test) dev_ctx.SetBlob(key_src_layout, p_src_layout);
```
Comment (tpatejko) [inline paddle/fluid/operators/activation_mkldnn_op.cc]: Are variables `p_src_data` and `p_src_layout` needed at inference? Are these shared pointers passed by value or by reference? You could potentially turn two if-statements in lines 126 and 128 into one, if these two shared pointers are passed by value to `SetBlob`.

File: paddle/fluid/operators/conv_mkldnn_op.cc [hunk 4]
```diff
@@ -536,12 +538,15 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
...
                        const mkldnn::engine& engine, const bool fuse_relu,
-                       const bool fuse_residual_conn) const {
+                       const bool fuse_residual_conn, bool is_test) const {
     memory::dims stride_dims = {strides[0], strides[1]};
     memory::dims padding_dims = {paddings[0], paddings[1]};
 
+    auto mkldnn_forward_prop_kind = is_test
+                                        ? mkldnn::prop_kind::forward_inference
+                                        : mkldnn::prop_kind::forward_training;
     auto conv_desc = mkldnn::convolution_forward::desc(
```
Comment (tpatejko) [inline paddle/fluid/operators/conv_mkldnn_op.cc:545]: Why is this check repeaded (line 512)? Can it be done once, in one place?

File: paddle/fluid/operators/pool_mkldnn_op.cc [hunk 5]
```diff
@@ -161,9 +156,20 @@ class PoolMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
       dev_ctx.SetBlob(key_pool_src_mem_p, src_memory);
       dev_ctx.SetBlob(key_pool_dst_mem_p, dst_memory);
 
-      pool_p = std::make_shared<pooling_forward>(*pool_pd, *(src_memory.get()),
-                                                 *(dst_memory.get()),
-                                                 *workspace_memory);
+      if (is_test) {
+        pool_p = std::make_shared<pooling_forward>(
+            *pool_pd, *(src_memory.get()), *(dst_memory.get()));
+      } else {
```
Comment (tpatejko) [inline paddle/fluid/operators/pool_mkldnn_op.cc:162]: Doesn't doing `*src_memory` have the same effect as `*(src_memory.get())`. If I'm not mistaken, `std::shared_ptr::operator*()` dereferences stored pointer, so it's the same as doing `*get()` on shared pointer.

File: paddle/fluid/framework/ir/is_test_pass.cc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+std::unique_ptr<ir::Graph> IsTestPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  VLOG(3) << "Sets is_test attrbiute to true and if it is missing, inserts it "
+             "for activations and pooling.";
+  std::array<std::string, 31> op_list = {
```
Comment (kbinias) [inline paddle/fluid/framework/ir/is_test_pass.cc]: Please change code to remove magic number

Comment (sfraczek) [reply paddle/fluid/framework/ir/is_test_pass.cc]: done

**Comment by sfraczek** [comment]
@tpatejko , regarding reorders, I've run resnet50 and the only reorder happens before softmax at the end. Is this how it should be?
```
mkldnn_verbose,exec,convolution,jit:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:x fdst:nChw16c,alg:convolution_direct,mb32_g1ic512oc512_ih7oh7kh3sh1dh0ph1_iw7ow7kw3sw1dw0pw1,39.439
mkldnn_verbose,exec,convolution,jit_1x1:avx512_common,forward_inference,fsrc:nChw16c fwei:OIhw16i16o fbia:x fdst:nChw16c,alg:convolution_direct,mb32_g1ic512oc2048_ih7oh7kh1sh1dh0ph0_iw7ow7kw1sw1dw0pw0,22.2341
mkldnn_verbose,exec,pooling,jit:avx512_common,forward_inference,fdata:nChw16c fws:undef,alg:pooling_avg_exclude_padding,mb32ic2048_ih7oh1kh7sh1ph0_iw7ow1kw7sw1pw0,1.88989
mkldnn_verbose,create,reorder,jit:uni,undef,in:f32_nChw16c out:f32_nchw,num:1,32x2048x1x1,0.129883
mkldnn_verbose,exec,reorder,jit:uni,undef,in:f32_nChw16c out:f32_nchw,num:1,32x2048x1x1,0.0480957
mkldnn_verbose,exec,softmax,ref:any,forward_inference,fdata:nc fdiff:undef,,mb32ic1000ih2075383169iw-32765,0.321045
Iteration: 3, accuracy: 0.812500, latency: 1.67097 s, fps: 19.150597
```
I will check for performance changes next.

File: paddle/fluid/operators/activation_mkldnn_op.cc (hunk 3 above)
Comment (sfraczek) [reply paddle/fluid/operators/activation_mkldnn_op.cc]: I didn't see any usage of them. I was actually hoping that a reviewer would verify if I didn't break things.
They are passed by value:
```
void MKLDNNDeviceContext::SetBlob(const std::string& name,           
                                  std::shared_ptr<void> data) const {
```
Ok I put both in one if().

File: paddle/fluid/operators/conv_mkldnn_op.cc (hunk 4 above)
Comment (sfraczek) [reply paddle/fluid/operators/conv_mkldnn_op.cc:545]: I can move it outside to line ~385

File: paddle/fluid/operators/pool_mkldnn_op.cc (hunk 5 above)
Comment (sfraczek) [reply paddle/fluid/operators/pool_mkldnn_op.cc:162]: Ok, I changed that.

**Comment by tpatejko** [comment]
@sfraczek From the snippet you posted, it looks fine.

**Comment by sfraczek** [comment]
I removed "virtual" keyword because of this:
```
cpplint..................................................................Failed
hookid: cpplint-cpp-source

paddle/fluid/inference/api/paddle_pass_builder.h:95:  "virtual" is redundant since function is already declared as "override"  [readability/inheritance] [4]
paddle/fluid/inference/api/paddle_pass_builder.h:127:  "virtual" is redundant since function is already declared as "override"  [readability/inheritance] [4]
Done processing paddle/fluid/inference/api/paddle_pass_builder.h
Total errors found: 2
```

**Comment by sfraczek** [comment]
Performance tests show slight improvement to speed after these changes. About 2.2% on batch one, processor SKX_6148.

**Comment by sfraczek** [comment]
Stack trace from CI
```
[11:39:25]192/439 Test #154: test_interpolate_op .............................***Exception: Other 98.33 sec
[11:39:25]terminate called after throwing an instance of 'paddle::platform::EnforceNotMet'
[11:39:25]  what():  cudaMemcpy failed in paddle::platform::GpuMemcpySync: an illegal memory access was encountered at [/paddle/paddle/fluid/platform/gpu_info.cc:155]
[11:39:25]PaddlePaddle Call Stacks: 
[11:39:25]0       0x7f5b653e5476p paddle::platform::EnforceNotMet::EnforceNotMet(std::__exception_ptr::exception_ptr, char const*, int) + 486
[11:39:25]1       0x7f5b66c915fap paddle::platform::GpuMemcpySync(void*, void const*, unsigned long, cudaMemcpyKind) + 234
[11:39:25]2       0x7f5b654314bdp _ZN6paddle6pybind7details18CastToPyBufferImplILb1ELm0EJfidlbhaNS_8platform7float16EEEclERKNS_9framework6TensorE + 637
[11:39:25]3       0x7f5b653dadc1p
[11:39:25]4       0x7f5b653f9776p pybind11_getbuffer + 150
[11:39:25]5       0x7f5bc3b78748p PyMemoryView_FromObject + 200
[11:39:25]6       0x7f5bbd9e1a03p
[11:39:25]7       0x7f5bbd9e4b78p
[11:39:25]8       0x7f5bbd9e5205p
[11:39:25]9       0x7f5bbda7c1f1p
[11:39:25]10      0x7f5bc3b790d9p PyCFunction_Call + 233
[11:39:25]11      0x7f5bc3c138fdp PyEval_EvalFrameEx + 38317
[11:39:25]12      0x7f5bc3c14366p
[11:39:25]13      0x7f5bc3c1098cp PyEval_EvalFrameEx + 26172
[11:39:25]14      0x7f5bc3c14366p
[11:39:25]15      0x7f5bc3c1098cp PyEval_EvalFrameEx + 26172
[11:39:25]16      0x7f5bc3c14366p
[11:39:25]17      0x7f5bc3c1098cp PyEval_EvalFrameEx + 26172
[11:39:25]18      0x7f5bc3c14366p
[11:39:25]19      0x7f5bc3c1098cp PyEval_EvalFrameEx + 26172
[11:39:25]20      0x7f5bc3c14366p
[11:39:25]21      0x7f5bc3c1098cp PyEval_EvalFrameEx + 26172
[11:39:25]22      0x7f5bc3c137c9p PyEval_EvalFrameEx + 38009
[11:39:25]23      0x7f5bc3c14366p
[11:39:25]24      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]25      0x7f5bc3b55d03p
[11:39:25]26      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]27      0x7f5bc3c0d70bp PyEval_EvalFrameEx + 13243
[11:39:25]28      0x7f5bc3c14366p
[11:39:25]29      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]30      0x7f5bc3b55c26p
[11:39:25]31      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]32      0x7f5bc3b3d4ddp
[11:39:25]33      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]34      0x7f5bc3b92521p
[11:39:25]35      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]36      0x7f5bc3c0d9bep PyEval_EvalFrameEx + 13934
[11:39:25]37      0x7f5bc3c14366p
[11:39:25]38      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]39      0x7f5bc3b55d03p
[11:39:25]40      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]41      0x7f5bc3c0d70bp PyEval_EvalFrameEx + 13243
[11:39:25]42      0x7f5bc3c14366p
[11:39:25]43      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]44      0x7f5bc3b55c26p
[11:39:25]45      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]46      0x7f5bc3b3d4ddp
[11:39:25]47      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]48      0x7f5bc3b92521p
[11:39:25]49      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]50      0x7f5bc3c0d9bep PyEval_EvalFrameEx + 13934
[11:39:25]51      0x7f5bc3c14366p
[11:39:25]52      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]53      0x7f5bc3b55d03p
[11:39:25]54      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]55      0x7f5bc3c0d70bp PyEval_EvalFrameEx + 13243
[11:39:25]56      0x7f5bc3c14366p
[11:39:25]57      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]58      0x7f5bc3b55c26p
[11:39:25]59      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]60      0x7f5bc3b3d4ddp
[11:39:25]61      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]62      0x7f5bc3b92521p
[11:39:25]63      0x7f5bc3b24c3ap PyObject_Call + 106
[11:39:25]64      0x7f5bc3c0d9bep PyEval_EvalFrameEx + 13934
[11:39:25]65      0x7f5bc3c137c9p PyEval_EvalFrameEx + 38009
[11:39:25]66      0x7f5bc3c137c9p PyEval_EvalFrameEx + 38009
[11:39:25]67      0x7f5bc3c14366p
[11:39:25]68      0x7f5bc3c14458p PyEval_EvalCodeEx + 72
[11:39:25]69      0x7f5bc3c1449bp PyEval_EvalCode + 59
[11:39:25]70      0x7f5bc3c45662p PyRun_FileExFlags + 178
[11:39:25]71      0x7f5bc3c457c7p PyRun_SimpleFileExFlags + 231
[11:39:25]72      0x7f5bc3c627c4p Py_Main + 3652
[11:39:25]73            0x400a54p main + 356
[11:39:25]74      0x7f5bc2c91d20p __libc_start_main + 256
[11:39:25]75            0x400b01p
[11:39:25]
[11:39:25]*** Aborted at 1542195564 (unix time) try "date -d @1542195564" if you are using GNU date ***
[11:39:25]PC: @                0x0 (unknown)
[11:39:25]*** SIGABRT (@0x418a) received by PID 16778 (TID 0x7f5bc41ec700) from PID 16778; stack trace: ***
[11:39:25]    @     0x7f5bc38a97e0 (unknown)
[11:39:25]    @     0x7f5bc2ca54f5 __GI_raise
[11:39:25]    @     0x7f5bc2ca6cd5 __GI_abort
[11:39:25]    @     0x7f5b76b87a8d __gnu_cxx::__verbose_terminate_handler()
[11:39:25]    @     0x7f5b76b85be6 (unknown)
[11:39:25]    @     0x7f5b76b85c13 std::terminate()
[11:39:25]    @     0x7f5b76b85d32 __cxa_throw
[11:39:25]    @     0x7f5b66c91618 paddle::platform::GpuMemcpySync()
[11:39:25]    @     0x7f5b654314bd paddle::pybind::details::CastToPyBufferImpl<>::operator()()
[11:39:25]    @     0x7f5b653dadc1 _ZZN8pybind116class_IN6paddle9framework9LoDTensorEINS2_6TensorEEE10def_bufferIZNS1_6pybindL13pybind11_initEvEUlRS4_E11_EERS5_OT_ENUlP7_objectPvE_4_FUNESE_SF_
[11:39:25]    @     0x7f5b653f9776 pybind11_getbuffer
[11:39:25]    @     0x7f5bc3b78748 PyMemoryView_FromObject
[11:39:25]    @     0x7f5bbd9e1a03 _array_from_buffer_3118
[11:39:25]    @     0x7f5bbd9e4b78 PyArray_FromAny
[11:39:25]    @     0x7f5bbd9e5205 PyArray_CheckFromAny
[11:39:25]    @     0x7f5bbda7c1f1 _array_fromobject
[11:39:25]    @     0x7f5bc3b790d9 PyCFunction_Call
[11:39:25]    @     0x7f5bc3c138fd PyEval_EvalFrameEx
[11:39:25]    @     0x7f5bc3c14366 (unknown)
[11:39:25]    @     0x7f5bc3c1098c PyEval_EvalFrameEx
[11:39:25]    @     0x7f5bc3c14366 (unknown)
[11:39:25]    @     0x7f5bc3c1098c PyEval_EvalFrameEx
[11:39:25]    @     0x7f5bc3c14366 (unknown)
[11:39:25]    @     0x7f5bc3c1098c PyEval_EvalFrameEx
[11:39:25]    @     0x7f5bc3c14366 (unknown)
[11:39:25]    @     0x7f5bc3c1098c PyEval_EvalFrameEx
[11:39:26]    @     0x7f5bc3c14366 (unknown)
[11:39:26]    @     0x7f5bc3c1098c PyEval_EvalFrameEx
[11:39:26]    @     0x7f5bc3c137c9 PyEval_EvalFrameEx
[11:39:26]    @     0x7f5bc3c14366 (unknown)
[11:39:26]    @     0x7f5bc3c14458 PyEval_EvalCodeEx
[11:39:26]    @     0x7f5bc3b55d03 (unknown)
[11:39:26]
```

**Comment by tpatejko** [comment]
LGTM.

**Review by tpatejko — APPROVED** [review]

**Pull request merged**